GDB: Added packet parser.

mrajwa opened the pull request:
This commit adds a *skeleton* of a parser for incoming GDB packets. 

Signed-off-by: Marcin Rajwa <marcin.rajwa@linux.intel.com>

jajanusz commented on src/include/sof/gdb/Makefile.am:
```diff
@@ -0,0 +1,20 @@
+noinst_LIBRARIES = libgdb.a
```
lib will be empty anyway. I guess content of whole file should be something like  `noinst_HEADERS =gdb.h ringbuffer.h`

lgirdwood commented on src/gdb/gdb.c:
```diff
@@ -0,0 +1,238 @@
+static unsigned char remcomOutBuffer[BUFMAX];
+
+/* local functions */
+static int getHex(unsigned char ch);
```
coding convention is get_hex, i.e. all lowercase for functions with _ to separate words.

mrajwa replied:
>One general question is that we have lots of while(1) {} loops reading or parsing GDB data. What happens if we have some corruption in the GDB command stream (or GDB unexpectedly quits), we could loop forever. We probably need to look at breaking out of the loops on error conditions.

Okay, this is good point actually, so let me elaborate a bit. 

**Corrupted data** - as you can see in code `/* TODO: handle wrong checksums */` there is no support for wrong checksum **on the gdb_stub side** at the moment. However, there is on the server side - it is tcp ip communication between GDB and gdb_server so the incoming data for gdb_stub on the FW side is quite consistent. 
BTW: I've been testing GDB for a weeks now,  guess how many corrupted packet I received? Zero. That's why I didn't bother to implement checksum validation on FW side yet.

**Infinite loop** - yes we constantly monitor incoming packets and once detected valid one we pass it further to parsing stage and this is what we want. What happens when GDB is hung or you lost connection? Then all you need to do is: restart the session and start over, gdb stub should still monitor incoming buffer and accept our requests.

Is that sufficient for you? I think at some point we should organize a video session showing GDB in action.

lgirdwood commented on src/gdb/gdb.c:
```diff
@@ -0,0 +1,238 @@
+
+/* main buffers */
+static unsigned char remcomInBuffer[BUFMAX];
+static unsigned char remcomOutBuffer[BUFMAX];
```
would it be worthwhile having local context (state and buffers) in a struct, that could in the future be placed by the linker out of harms way in the memory map (and in a platform consistent location).

mrajwa replied:
We can put it into a GDB related .data section anyway. To be honest I don't see any benefit of keeping these few variables inside a struct. A struct is useful when you have related data and need to pass them around or make objects/other constructs on their basics - here it will only add syntax complexity IMHO.

mrajwa replied:
BTW: Your suggestion about keeping GDB related stuff in dedicated section will be useful later on when we would like to load a module on demand.

lgirdwood commented on src/gdb/gdb.c:
```diff
@@ -0,0 +1,238 @@
+
+/* registers backup conatiners */
+int sregs[256];
+int aregs[64];
```
The naming here looks xtensa specific, should these also be "unsigned long" to align with target register size.

mrajwa replied:
Our target registers are 4 bytes, and both int and long are 4 bytes so it is all fine.

mmaka1 replied:
@mrajwa Should the magic numbers like 64 be replaced with XCHAL_NUM_AREGS etc. defined by the arch spec?

mrajwa replied:
@mmaka1, yes I will replace it in following commits.

lgirdwood commented on src/gdb/gdb.c:
```diff
@@ -0,0 +1,238 @@
+
+static const char hexchars[] = "0123456789abcdef";
+
+
```
Just use one new line to separate functions/blocks.

lgirdwood commented on src/gdb/gdb.c:
```diff
@@ -0,0 +1,238 @@
+			putDebugChar('+');
+
+		/* if a sequence char is present, reply the sequence ID */
+			if (buffer[2] == ':') {
```
indentation does not match comment indent.

lgirdwood commented on src/gdb/gdb.c:
```diff
@@ -0,0 +1,238 @@
+		unsigned char command = *request++;
+
+		switch (command) {
+		case 'c':
```
Can you comment what each command does. I'm assuming c is "continue" iirc.

lgirdwood commented on src/gdb/gdb.c:
```diff
@@ -0,0 +1,238 @@
+static int hexToInt(unsigned char **ptr, int *intValue)
+{
+	int numChars = 0;
+	int hexValue;
```
naming convention applies to variable names too e.g. hex_value

lgirdwood commented on src/gdb/gdb.c:
```diff
@@ -0,0 +1,238 @@
+		numChars++;
+		(*ptr)++;
+	}
+	return (numChars);
```
no need for () around return value.

lgirdwood commented on src/gdb/ringbuffer.c:
```diff
@@ -0,0 +1,41 @@
+
+volatile struct ring * const rx = (void *) 0x9e008000;
+volatile struct ring * const tx = (void *) 0x9e008120;
+volatile struct ring * const debug = (void *) 0x9e008220;
```
These addresses should be defined in memory.h

mrajwa replied:
@lgirdwood, they belong to GDB and only to GDB - meaning they will not be used by any other component other than GDB. So does it really make sense to move it outside GDB ? If a month from now we decided to load GDB on demand these addresses in memory.h will must be again taken away or will become a residue.

lgirdwood replied:
ok, so how do these addresses align with the xtensa memory map ? If they are special debug only addresses should they not be in core-isa.h ? The intention for memory.h is to be like core-isa.h and allow differentiation per platform and architecture, so this would break on non xtensa architectures.

mrajwa replied:
I rethinked this case once more and we have two choices in my opinion - either to keep these addresses where they are now with `#if defined` for different platform (because current version has only debug addresses for ICL) or move them to each platform memory.h file with `#if defined CONFIG_GDB_DEBUG` (we can even skip this precompiler check with no harm). The advantage of having it defined inside GDB rather than memory.h is: less files to maintain, better encapsulation and modular approach.

lgirdwood commented on src/include/sof/gdb/gdb.h:
```diff
@@ -0,0 +1,12 @@
+#ifndef GDB_H
+#define GDB_H
+
+#define BUFMAX 256
```
We should prefix these macros with GDB_

lgirdwood commented on src/include/sof/gdb/gdb.h:
```diff
@@ -0,0 +1,12 @@
+void gdbInit(void);
+void handleException(void);
+extern void debug_exception(unsigned char *str);
+extern void initDebugException(void);
```
likewise prefix these APIS with gdb_

lgirdwood commented on src/include/sof/gdb/ringbuffer.h:
```diff
@@ -0,0 +1,40 @@
+
+#define RING_SIZE 128
+#define DEBUG_RX_BASE (mailbox_get_debug_base())
+#define DEBUG_TX_BASE (mailbox_get_debug_base() + 0x100)
```
ditto for prefixing GDB_ and gdb_ here with these public APIS and macros.

mrajwa commented:
>One general question is that we have lots of while(1) {} loops reading or parsing GDB data. What happens if we have some corruption in the GDB command stream (or GDB unexpectedly quits), we could loop forever. We probably need to look at breaking out of the loops on error conditions.

Okay, this is good point actually, so let me elaborate a bit. 

**Corrupted data** - as you can see in code `/* TODO: handle wrong checksums */` there is no support for wrong checksum **on the gdb_stub side** at the moment. However, there is on the server side - it is tcp ip communication between GDB and gdb_server so the incoming data for gdb_stub on the FW side is quite consistent. 
BTW: I've been testing GDB for a weeks now,  guess how many corrupted packet I received? Zero. That's why I didn't bother to implement checksum validation on FW side yet.

**Infinite loop** - yes we constantly monitor incoming packets and once detected valid one we pass it further to parsing stage and this is what we want. What happens when GDB is hung or you lost connection? Then all you need to do is: restart the session and start over, gdb stub should still monitor incoming buffer and accept our requests.

Is that sufficient for you? I think at some point we should organize a video session showing GDB in action.

lgirdwood commented:
@mrajwa that sounds fine to me now that you have explained the detail and re-connecting GDB re-establishes the cmd stream

mrajwa commented:
@lgirdwood I updated the software, also in the meantime I adjusted build for new directories layout and generalized debug addresses to work on base of memory.h

mrajwa commented:
@lgirdwood ping
